ENH: cythonized spherical Voronoi region polygon vertex sorting

tylerjereddy opened the pull request:
### Purpose
The objective is to increase the speed of the polygon (spherical Voronoi region) vertex sorting code used by `SphericalVoronoi.sort_vertices_of_regions()` by cythonizing the pure Python implementation.

### Initial Results
Benchmarking against the tip of master on my machine:
`asv continuous --bench SphericalVorSort b48f233299116226a4a2f4648dcddd975c2725ba ba32cf965e8336556a1a8eac3d79cae99ef30093`

```
· Creating environments
· Discovering benchmarks
·· Uninstalling from virtualenv-py2.7-Cython0.22-Tempita0.5.2-numpy1.8.2-six.
·· Installing into virtualenv-py2.7-Cython0.22-Tempita0.5.2-numpy1.8.2-six..
· Running 2 total benchmarks (2 commits * 1 environments * 1 benchmarks)
[  0.00%] · For scipy commit hash ba32cf96:
[  0.00%] ·· Building for virtualenv-py2.7-Cython0.22-Tempita0.5.2-numpy1.8.2-six.........................................................................................................................................................................
[  0.00%] ·· Benchmarking virtualenv-py2.7-Cython0.22-Tempita0.5.2-numpy1.8.2-six
[ 50.00%] ··· Running spatial.SphericalVorSort.time_spherical_polygon_vertex_sorting                                                                                                            92.25μs;...
[ 50.00%] · For scipy commit hash b48f2332:
[ 50.00%] ·· Building for virtualenv-py2.7-Cython0.22-Tempita0.5.2-numpy1.8.2-six....................................................................................................................................................................
[ 50.00%] ·· Benchmarking virtualenv-py2.7-Cython0.22-Tempita0.5.2-numpy1.8.2-six
[100.00%] ··· Running spatial.SphericalVorSort.time_spherical_polygon_vertex_sorting                                                                                                           797.58μs;...    
    before     after       ratio
  [b48f2332] [ba32cf96]
-  797.58μs    92.25μs      0.12  spatial.SphericalVorSort.time_spherical_polygon_vertex_sorting(10)
-     1.04s    96.45ms      0.09  spatial.SphericalVorSort.time_spherical_polygon_vertex_sorting(10000)
-  544.51ms    47.08ms      0.09  spatial.SphericalVorSort.time_spherical_polygon_vertex_sorting(5000)
-  125.51ms     8.40ms      0.07  spatial.SphericalVorSort.time_spherical_polygon_vertex_sorting(1000)
-   13.64ms   870.80μs      0.06  spatial.SphericalVorSort.time_spherical_polygon_vertex_sorting(100)

SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.
```
-------------
`line_profiler` result for cython function:
```
Timer unit: 1e-06 s

Total time: 0.154271 s
File: /Users/treddy/.ipython/cython/_cython_magic_f8b3bd94f0b62b2e37e528dd3bc55e1d.pyx
Function: sort_vertices_of_regions at line 15

Line #      Hits         Time  Per Hit   % Time  Line Contents
==============================================================
    15                                           def sort_vertices_of_regions(int[:,::1] simplices, regions):
    16                                             cdef unsigned int n, k, s, i
    17         1            4      4.0      0.0    cdef int num_regions = len(regions)
    18                                             cdef int current_simplex, current_vertex, remaining_count
    19                                             cdef int cs_identified, remaining_size
    20                                             cdef long[:] remaining
    21      2002          978      0.5      0.6    cdef long[:] sorted_vertices = np.zeros(max([len(region) for region in
    22         2            7      3.5      0.0  	  regions]), dtype = np.int64)
    23                                           
    24         1            5      5.0      0.0    for n in range(num_regions):
    25      2000          773      0.4      0.5      remaining_count = 0
    26      2000        15808      7.9     10.2      remaining = np.asarray(regions[n][:])
    27      2000          723      0.4      0.5      remaining_size = remaining.shape[0]
    28      2000          722      0.4      0.5      sorted_vertices[:] = -2
    29      2000          706      0.4      0.5      current_simplex = remaining[0]
    30      2000          713      0.4      0.5      for i in range(3):
    31      3271         1298      0.4      0.8        k = simplices[current_simplex][i]
    32      3271         1133      0.3      0.7        if k != n:
    33      2000          684      0.3      0.4          current_vertex = k
    34      2000          716      0.4      0.5          break
    35      2000          708      0.4      0.5      sorted_vertices[remaining_count] = current_simplex
    36      2000          665      0.3      0.4      remaining_count += 1
    37      2000         1994      1.0      1.3      remaining_filter(remaining.shape[0], remaining, current_simplex)
    38      2000          740      0.4      0.5      while remaining_size > remaining_count:
    39      9988         3277      0.3      2.1        cs_identified = 0
    40      9988         3492      0.3      2.3        for i in range(remaining_size):
    41     41730        14471      0.3      9.4          if remaining[i] == -2:
    42     31742        10729      0.3      7.0            continue
    43      9988         3402      0.3      2.2          s = remaining[i]
    44      9988         3334      0.3      2.2          for j in range(3):
    45     19033         7036      0.4      4.6            if current_vertex == simplices[s,j]:
    46      9988         3375      0.3      2.2              current_simplex = remaining[i]
    47      9988         3391      0.3      2.2              cs_identified += 1
    48      9988         3386      0.3      2.2              break
    49      9988         3424      0.3      2.2          if cs_identified > 0:
    50      9988         3394      0.3      2.2            break
    51      9988         3389      0.3      2.2        for i in range(3):
    52     20007         6778      0.3      4.4          s = simplices[current_simplex, i]
    53     20007         7040      0.4      4.6          if s != n and s != current_vertex:
    54      9988         3232      0.3      2.1            current_vertex = s
    55      9988         3373      0.3      2.2            break
    56      9988         3550      0.4      2.3        sorted_vertices[remaining_count] = current_simplex
    57      9988         3243      0.3      2.1        remaining_count += 1
    58      9988         9806      1.0      6.4        remaining_filter(remaining.shape[0], remaining, current_simplex)
    59      2000        14468      7.2      9.4      regions_arr = np.asarray(sorted_vertices)
    60      2000         8304      4.2      5.4      regions[n] = regions_arr[regions_arr > -2].tolist()
```

### To-do list:

- [x] Add appropriate header info / versionadded / changelog type stuff if we decide to incorporate some form of this optimization
- [x] Consider renaming `_sv_sort_vertices.pyx` to something more general so that the spherical Voronoi code proper may eventually be cythonized in the same file without bloating `spatial` with cython source files? or maybe use a subfolder?
- [ ] Look for other optimization opportunities? There's a bit of awkwardness with [memoryview coercion to numpy arrays](http://cython.readthedocs.io/en/latest/src/userguide/memoryviews.html#coercion-to-numpy). My basic idea was to pre-allocate arrays as much as possible and use filter values (-2 in my code) to accommodate cases where smaller arrays are used in the control flow.

### Why has a new unit test been added with this enhancement?
Because the only unit test that currently verifies vertex sorting behaviour did not seem to be sufficiently sensitive to detect certain issues. In particular, when the nested lists had matching lengths but different contents, the original test could pass in some instances. I suspect this relates to the heterogeneity (`dtype=object`) for nested list data structures although quick interactive testing suggests that `np.testing.assert_array_equal()` does ok with these cases too. In any case, an additional test where the data has been flattened  was added (the old test was retained too), as this would always find an issue during the development process.

rgommers commented:
Significant speedup and the code is not a whole lot longer, so seems like a good idea to me.

Agreed that the file name is too specific. How about simply ``_voronoi.pyx``? Or ``_voronoi_utils.pyx`` if you want something a little more descriptive.

rgommers commented:
I'm a bit worried by the ``int``s and ``long``s in this code. Looks like some of those could overflow. Make those all ``npy_intp``?

rgommers commented:
@tylerjereddy just in case you haven't seen it, this is a good reference for integer types: https://github.com/scikit-learn/scikit-learn/wiki/C-integer-types:-the-missing-manual

tylerjereddy commented:
With one exception, I've replaced all usage of `int` with `np.npy_intp`. I just thought I'd play it safe, though we do take a minor performance hit for the adjusted type casting at lower data sizes. Looks like we're still comfortably an order of magnitude faster for larger data sizes though. These benchmarks are new Cython vs. original Cython (not comparing back to the old Python source).

```
   before     after       ratio
  [ba32cf96] [02dc772b]
+   98.53μs   116.44μs      1.18  spatial.SphericalVorSort.time_spherical_polygon_vertex_sorting(10)
+  869.07μs   957.89μs      1.10  spatial.SphericalVorSort.time_spherical_polygon_vertex_sorting(100)

SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.
```
Attempts to cast the final `int` as the `np.npy_intp` type result in `ValueError: Buffer dtype mismatch, expected 'npy_intp' but got 'int'`. This is where the Python data is read into the Cython function `sort_vertices_of_regions()`. Not sure if there is still overflow danger there? The memoryview object `simplices` that is typed as `int` would (I suppose) have overflow issues if there are enough input data points to have indices that exceed the max value tolerated by the `int` type. If I can't cast the type at the function def level, the only other option would be to set the type in the parent python module?! If we do that, we're basically saying that the data type returned by `scipy.spatial.ConvexHull.simplices` is not sufficiently large to store what we need?

I suppose even better would be if we could write a robust unit test that probes for int overflow susceptibility on the Cython function, but I don't see much online about doing this in a straightforward way.

I followed one of your suggestions for renaming the Cython source file.

rgommers commented:
> I suppose even better would be if we could write a robust unit test that probes for int overflow susceptibility on the Cython function, but I don't see much online about doing this in a straightforward way.

That's not really possible. Such failures depend on the input data. It's not always easy to predict which ``int``s can get values > 2**31 for some input parameters. We just fixed one in cluster that only showed the issue when the input linkage matrix was such that you allocated 16GB, which is hard to test.

rgommers commented on scipy/spatial/_voronoi.pyx:
```diff
@@ -0,0 +1,60 @@
+    
+@cython.boundscheck(False)
+def sort_vertices_of_regions(int[:,::1] simplices, regions):
+  cdef np.npy_intp n, k, s, i
```
indentation is 2 spaces here, and 3 spaces in the function above. can you make it 4 everywhere?

rgommers commented:
> Attempts to cast the final int as the np.npy_intp type result in ValueError: Buffer dtype mismatch, expected 'npy_intp' but got 'int'. This is where the Python data is read into the Cython function sort_vertices_of_regions().

The first question there is whether the code works. Is it possible that ``simplices`` contains ``int64`` on a 32-bit system or ``int32`` on a 64-bit system? ``ConvexHull`` says it gives ``ndarray of ints``, but I haven't checked any further than that docstring.

tylerjereddy commented:
@rgommers Just dealing with the most complicated issue first -- tracking the data type for `ConvexHull` simplices:

- `self.simplices` is obtained as the first argument of the `tuple` returned by `qhull.get_simplex_facet_array() `
- within `get_simplex_facet_array()`, this object corresponds to the `facets` object
- `facets` is initially typed as `cdef np.ndarray[np.npy_int, ndim=2] facets`
- `facets` is later redefined with `facets = np.zeros((j, facet_ndim), dtype=np.intc)`
- `facets` is then filled in a set of cython for loops and returned without any other apparent type changes

Testing interactively in scipy `0.19.0.dev0` on my 64-bit laptop:

```python
import numpy as np
type(np.intc(1))
# numpy.int32
import scipy
from scipy.spatial import ConvexHull
hull = ConvexHull(np.random.random_sample((5,3)))
hull.simplices
#array([[1, 0, 2],
#       [1, 3, 2],
#       [4, 0, 2],
#       [4, 3, 2],
#       [4, 1, 0],
#       [4, 1, 3]], dtype=int32)
```

Are you proposing that we use `np.npy_intp` for `facets` (and therefore `simplices`) in `ConvexHull`?

tylerjereddy commented:
- Indentation of the cython file should be more consistent & 4 spaces now
- Fixed a few more `flake8` type minor issues in the cython file
- Added minor documentation-related things (cython file header and short line in release notes)

rgommers commented:
> Are you proposing that we use `np.npy_intp` for `facets` (and therefore `simplices`) in `ConvexHull`?

Not sure, it may be fine as is. I'm not familiar enough with this code - looks like the type is fixed so that's good. So if there's not an issue with using 32-bit ints for this attribute (would you ever have elements > 2**31 in there?) then leave as is I guess.

rgommers commented:
Can you add ``scipy/spatial/_voronoi.c`` to ``.gitignore``?

pv commented:
Re: intp --- this probably should match the type qhull uses, which I 
think is int.

tylerjereddy commented:
For context, the maximum integer value in `simplices` is constrained by the number of input points, so if `ConvexHull` can run properly with a number of input points that exceeds the 32-bit max int size on a machine, we'd have an issue.

So, testing...
 
```python
import numpy as np
import scipy
from scipy.spatial import ConvexHull

ii32 = np.iinfo(np.int32)
points = np.random.random_sample((ii32.max, 2))
ConvexHull(points)
```

Execute the above on:

1) 64-bit machine with 16 GM RAM and scipy `0.19-dev`:
```
Traceback (most recent call last):
  File "convex_hull_tester.py", line 7, in <module>
    ConvexHull(points)
  File "scipy/spatial/qhull.pyx", line 2383, in scipy.spatial.qhull.ConvexHull.__init__ (scipy/spatial/qhull.c:23023)
  File "scipy/spatial/qhull.pyx", line 434, in scipy.spatial.qhull._Qhull.__init__ (scipy/spatial/qhull.c:5541)
scipy.spatial.qhull.QhullError: QH6013 qhull input error: input is less than 2-dimensional since it has the same x coordinate

While executing:  | qhull i Qt
Options selected for Qhull 2015.2.r 2016/01/18:
  run-id 1962850242  incidence  Qtriangulate  _pre-merge  _zero-centrum
  _max-width  0  Error-roundoff 7.4e-16  _one-merge 3.7e-15
  _near-inside 1.9e-14  Visible-distance 1.5e-15  U-coplanar-distance 1.5e-15
  Width-outside 3e-15  _wide-facet 8.9e-15

```
2) 64-bit supercomputer node with 128 GB RAM and scipy `0.18`:
```
QH6013 qhull input error: input is less than 2-dimensional since it has the same x coordinate

While executing:  | qhull i Qt
Options selected for Qhull 2012.1 2012/02/18:
  run-id 1963774628  incidence  Qtriangulate  _pre-merge  _zero-centrum
  _max-width  0  Error-roundoff 4.3e-16  _one-merge 2.2e-15
  _near-inside 1.1e-14  Visible-distance 8.7e-16  U-coplanar-distance 8.7e-16
  Width-outside 1.7e-15  _wide-facet 5.2e-15
Traceback (most recent call last):
  File "convex_hull_tester.py", line 7, in <module>
    ConvexHull(points)
  File "scipy/spatial/qhull.pyx", line 2266, in scipy.spatial.qhull.ConvexHull.__init__ (scipy/spatial/qhull.c:21663)
  File "scipy/spatial/qhull.pyx", line 338, in scipy.spatial.qhull._Qhull.__init__ (scipy/spatial/qhull.c:3879)
scipy.spatial.qhull.QhullError: Qhull error

```
Try gradually reducing number of random input points to `ConvexHull`:

| Points | 16 GB machine (`0.19 dev`) | 128 GB machine (`0.18`) |
|---------|----------------------|---------------------|
| `ii32.max - 1` | fail as above | fail as above |
|`int(ii32.max / 10.)`|ok | ok|
|`int(ii32.max / 2.)`|`Bus error: 10` | `Segmentation fault (core dumped)`|
|`int(ii32.max / 3.)`|`Segmentation fault: 11` | `QH6235 qhull error (qh_memalloc): negative request size (-477216976).  Did int overflow due to high-D? `|
|`int(ii32.max / 5.)`|`QH6235 qhull error (qh_memalloc): negative request size (-858993448).  Did int overflow due to high-D?` | `QH6235 qhull error (qh_memalloc): negative request size (-858993448).  Did int overflow due to high-D?`|
|`int(ii32.max / 7.)`|`QH6235 qhull error (qh_memalloc): negative request size (-1840700256).  Did int overflow due to high-D?` | `QH6235 qhull error (qh_memalloc): negative request size (-1840700256).  Did int overflow due to high-D?`|
|`int(ii32.max / 9.)`| ok  | ok  |

For the largest successfully-tested case:
```python
points = np.random.random_sample((int(ii32.max / 9.), 2))
hull = ConvexHull(points)
max_simplices = np.amax(hull.simplices)
print 'points.shape:', points.shape
print 'max_simplices:', max_simplices
```
yields:
``` 
points.shape: (238609294, 2)
max_simplices: 234004476
```
Here, `max_simplices` is a bit smaller than the number of points in the input because not all input points will necessarily be part of the convex hull (elastic band surrounding the point set).

The above inputs are far larger than I would normally employ in research & expect a reasonable chance of success for. Overall, I suspect you'll agree that there's little danger that `simplices` will contain an integer that overflows because `ConvexHull` does not appear to be able to handle input large enough to exceed the 32-bit limit for the number of rows (or points, and therefore the indices represented by `ints` in `simplices`).

So, it looks like the "limitation" is at the qhull level? We could look into the quickhull algorithm publication to see if the space complexity is mentioned & consistent with the limits shown here, though one would think that this would have been checked by the qhull team.

tylerjereddy commented:
- rebased against master

rgommers commented:
> Overall, I suspect you'll agree that there's little danger that `simplices` will contain an integer that overflows because `ConvexHull` does not appear to be able to handle input large enough to exceed the 32-bit limit for the number of rows

Indeed, I agree. Nice test work. The segmentation faults would be good to avoid, but that's a separate issue from this PR. Maybe open a new issue?

rgommers commented:
Made one minor comment, but overall LGTM.

tylerjereddy commented:
- I've used the Cython equivalent of a named constant as you suggest (this is apparently done via an anonymous `enum` declaration as Cython has no special syntax for defining a constant -- see: [type defs](https://cython.readthedocs.io/en/latest/src/userguide/language_basics.html#c-variable-and-type-definitions))
- repeated benchmarks against local tip of master to check for any substantial regressions after the various fixes:
```
   before     after       ratio
  [b6701e04] [d57bc69d]
-  845.60μs    97.52μs      0.12  spatial.SphericalVorSort.time_spherical_polygon_vertex_sorting(10)
-     1.08s   106.73ms      0.10  spatial.SphericalVorSort.time_spherical_polygon_vertex_sorting(10000)
-   10.49ms   941.68μs      0.09  spatial.SphericalVorSort.time_spherical_polygon_vertex_sorting(100)
-  556.31ms    49.22ms      0.09  spatial.SphericalVorSort.time_spherical_polygon_vertex_sorting(5000)
-  109.19ms     9.18ms      0.08  spatial.SphericalVorSort.time_spherical_polygon_vertex_sorting(1000)
```
Looks about the same compared to the benchmarks when the PR was opened.

person142 commented:
> this is apparently done via an anonymous `enum` declaration as Cython has no special syntax for defining a constant

`DEF array_filler = -2`

tylerjereddy commented:
@person142 

So the [Cython 0.25 docs](https://cython.readthedocs.io/en/latest/src/userguide/language_basics.html#c-variable-and-type-definitions) are out of date?


<img width="911" alt="untitled" src="https://cloud.githubusercontent.com/assets/7903078/20663994/60b0411e-b551-11e6-9d5e-ea13babcad47.png">

person142 commented:
I remember reading exactly that line a while back (well, older version of Cython) and thinking it wasn't possible too. So... I guess so. See e.g.

https://github.com/scipy/scipy/blob/master/scipy/special/_loggamma.pxd#L31

for an example in the wild. (Or am I misunderstanding what you want to do?)

person142 commented:
Unless in the docs they mean constant in the sense of `const` in C--that's still not possible AFAIK. But you can use `DEF` to get something that is resolved at compile time--i.e. the equivalent of `#define`ing a value in C.

tylerjereddy commented:
Ok, good to know. I'm not clear on whether there's any functional difference between the two approaches, but since you've already used one approach in the `scipy` codebase I should perhaps switch to `DEF` with an upper-case constant name rather than the slightly-obscure anonymous `enum` thing I've used.

tylerjereddy commented:
- I've switched the named constant to the format already used in the `scipy` codebase. No substantial change in benchmark results on my machine.

rgommers commented:
@tylerjereddy do you still want to look at that third checkbox of your todo list? I'm tempted to hit the green button here.

tylerjereddy commented:
@rgommers Looks good to me. My notes indicate that I had addressed all reviewer comments / concerns as of Nov. 29/ 2016.

rgommers commented:
Allright, in it goes. Thanks Tyler!
